test/lib/validation/form/no-required-notes.spec.js

```diff
@@ -0,0 +1,141 @@
+      const req = required ? `required="${required}"` : '';
+      return `<bind nodeset="${name}" type="${type}" ${calc} ${read} ${req}/>`;
+    })
+    .join('');
+
+const getXmlDoc = (fields, instance) => domParser.parseFromString(getXml(createBindData(fields), instance));
+const xformPath = '/my/form/path/form.xml';
+
+const assertEmpty = (output) => {
+  expect(output.warnings).is.empty;
+  expect(output.errors, output.errors).is.empty;
+};
+
+const getExpectedErrorMsg = requiredNotes => `Form at ${xformPath} contains the following note fields with 'required' expressions: [${requiredNotes.join(', ')}]`;
+
+describe('no-required-notes', () => {
+  it('resolves OK for form with no notes', () => {
+    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc([]) })
+      .then(output => assertEmpty(output));
+  });
+
+  it('resolves OK for note that is not required', () => {
+    const fields = [{
+      name: '/data/name',
+      type: 'string',
+      readonly: 'true()',
+      required: 'false()'
+    },{
+      name: '/data/address/street-nbr',
+      type: 'string',
+      readonly: 'true()'
+    }];
+    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc(fields) })
+      .then(output => assertEmpty(output));
+  });
+
+  it('resolves OK for calculate that is required', () => {
+    const fields = [{
+      name: '/data/name',
+      type: 'string',
+      calculate: 'concat("Hello", "World")',
+      readonly: 'true()',
+      required: 'true()'
+    }];
+    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc(fields) })
+      .then(output => assertEmpty(output));
+  });
+
+  it('resolves OK for string question that is not readonly', () => {
+    const fields = [{
+      name: '/data/name',
+      type: 'string',
+      readonly: 'false()',
+      required: 'true()'
+    },{
+      name: '/data/address/street-nbr',
+      type: 'string',
+      readonly: 'false()'
+    }];
+    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc(fields) })
+      .then(output => assertEmpty(output));
+  });
+
+  it('resolves OK for non-string question', () => {
+    const fields = [{
+      name: '/data/age',
+      type: 'int',
+      readonly: 'true()',
+      required: 'true()'
+    }];
+    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc(fields) })
+      .then(output => assertEmpty(output));
+  });
+
+  it('returns errors for required notes', () => {
+    const fields = [{
+      name: '/data/name',
+      type: 'string',
+      readonly: 'true()',
+      required: 'true()'
+    },{
+      name: '/data/address/street-nbr',
+      type: 'string',
+      readonly: 'true()',
+      calculate: '',
+      required: '/data/age > 5'
+    }];
+    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc(fields) })
+      .then(output => {
+        expect(output.warnings).is.empty;
+        expect(output.errors).to.deep.equal([getExpectedErrorMsg(fields.map(f => f.name))]);
```

Comment:
Please does the `getExpectedErrorMsg` function returns the expected error messages or just all the required fields ?
I added a non-string question one to the fields
```
  it('returns errors for required notes', () => {
    const fields = [{
      name: '/data/name',
      type: 'string',
      readonly: 'true()',
      required: 'true()'
    },{
      name: '/data/address/street-nbr',
      type: 'string',
      readonly: 'true()',
      calculate: '',
      required: '/data/age > 5'
    },
    {
      name: '/data/age',
      type: 'int',
      readonly: 'true()',
      required: 'true()'
    }];
    return noRequiredNotes.execute({ xformPath, xmlDoc: getXmlDoc(fields) })
      .then(output => {
        expect(output.warnings).is.empty;
        console.log(getExpectedErrorMsg(fields.map(f => f.name)));
        console.log(output.errors);
        expect(output.errors).to.deep.equal([getExpectedErrorMsg(fields.map(f => f.name))]);
      });
  });
```
and then the test failed probably because `getExpectedErrorMsg` only return the required fields. 
** Test Output : **
```
Form at /my/form/path/form.xml contains the following note fields with 'required' expressions: [/data/name, /data/address/street-nbr, /data/age]
[
  "Form at /my/form/path/form.xml contains the following note fields with 'required' expressions: [/data/name, /data/address/street-nbr]"
]

AssertionError: expected [ Array(1) ] to deeply equal [ Array(1) ]
```

Reply:
So, the [`getExpectedErrorMsg`](https://github.com/medic/cht-conf/blob/393ba81c47c20cacc48c059e7de9e2fa699207a7/test/lib/validation/form/no-required-notes.spec.js#L61) just returns a string that contains the the given field names that were provided as the argument.  It just so happens that for this test both of the entries in `fields` are required notes (which should be included in the error message). However, if you add another field that should not be included in the error, you would need to filter that field out before passing the array of field names to `getExpectedErrorMsg`.

Reply:
ok, makes sense ! 
LGTM